docs/USAGE-GUIDE.md

```diff
@@ -607,66 +607,43 @@ In TLS1.3, session ticket messages are sent after the handshake as "post-handsha
 
-- **S2N_CLIENT_HELLO_CB_NONBLOCKING**:
+`s2n_client_hello_get_cipher_suites()` will retrieve the list of cipher suites sent by the client.
 
-    In non-blocking mode, s2n-tls expects the callback to not complete its work. If the callback
-    returns a response code of 0 s2n-tls will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED**
-    error type and **s2n_blocked_status** set to **S2N_BLOCKED_ON_APPLICATION_INPUT**.
-    The handshake is paused and further calls to **s2n_negotiate** will continue to return the
-    same error until **s2n_client_hello_cb_done** is invoked for the **s2n_connection** to resume
-    the handshake. This allows s2n-tls clients to process client_hello without
-    blocking and then resume the handshake at a later time.
-    If any of the connection properties were changed on the basis of the server_name extension then
-    **s2n_connection_server_name_extension_used** must be invoked before marking the callback done.
+`s2n_client_hello_get_session_id()` will get the session ID sent by the client in the ClientHello message. Note that this value may not be the session ID eventually associated with this particular connection since the session ID can change when the server sends the Server Hello. The official session ID can be retrieved with `s2n_connection_get_session_id()`.
 
-The callback can return a negative value to make s2n-tls terminate the
-handshake early with a fatal handshake failure alert.
+Call `s2n_client_hello_get_extensions()` to retrieve the entire list of extensions sent in the Client Hello. Calling `s2n_client_hello_get_extension_by_id()` allows you to interrogate the `s2n_client_hello` struct for a specific extension.
 
-```c
-int s2n_client_hello_cb_done(struct s2n_connection *conn)
-```
-Marks the non-blocking callback as complete.
-Can be invoked from within the callback when operating in non-blocking mode
-to continue the handshake.
+### Client Hello Callback
 
-```c
-int s2n_client_server_name_used(struct s2n_connection *conn)
-```
-Indicates that connection properties were changed on the basis of server_name.
-Triggers a s2n-tls server to send the server_name extension. Must be called
-before s2n-tls finishes processing the ClientHello.
+Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully as modifying the connection mid-handshake can be dangerous. In particular, switching from a `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack. 
+
```

Comment:
I'm now wondering if "dangerous" is too harsh. Most things you do are safe, there are just things you can do that are unsafe. Maybe if we're more specific about what exactly is dangerous?
```suggestion
Users can access the Client Hello during the handshake by setting the callback `s2n_config_set_client_hello_cb()`. A possible use-case for this is to modify the `s2n_connection` based on information in the Client Hello. This should be done carefully, as modifying the connection in response to untrusted input can be dangerous. In particular, switching from an `s2n_config` that supports TLS1.3 to one that does not opens the server up to a possible version downgrade attack. 

```